src/library/scala/collection/mutable/HashMap.scala

```diff
@@ -107,6 +107,55 @@ class HashMap[K, V](initialCapacity: Int, loadFactor: Double)
     }
   }
 
+  // Override updateWith for performance, so we can do the update while hashing
+  // the input key only once and performing one lookup into the hash table
+  override def updateWith(key: K)(remappingFunction: Option[V] => Option[V]): Option[V] = {
+    val hash = computeHash(key)
+    val indexedHash = index(hash)
+
+    var foundNode: Node[K, V] = null
+    var previousNode: Node[K, V] = null
+    table(indexedHash) match {
+      case null =>
+      case nd =>
+        @tailrec
+        def findNode(prev: Node[K, V], nd: Node[K, V], k: K, h: Int): Unit = {
+          if(h == nd.hash && k == nd.key) {
+            previousNode = prev
+            foundNode = nd
+          }
+          else if((nd.next eq null) || (nd.hash > h)) ()
+          else findNode(nd, nd.next, k, h)
+        }
+
+        findNode(null, nd, key, hash)
+    }
+
+    val previousValue = foundNode match {
+      case null => None
+      case nd => Some(nd.value)
+    }
+
+    val nextValue = remappingFunction(previousValue)
+
+    (previousValue, nextValue) match {
+      case (None, None) => // do nothing
+
+      case (Some(_), None) =>
+        if (previousNode != null) previousNode.next = foundNode.next
+        else table(indexedHash) = foundNode.next
+        contentSize -= 1
+
+      case (None, Some(value)) =>
+        if(contentSize + 1 >= threshold) growTable(table.length * 2)
+        table(indexedHash) = new Node[K, V](key, hash, value, table(indexedHash))
```

Comment:
I think there needs to be some logic in here to insert the new node in the correct place in linked-list. @szeiger recently made a change to mutable.HashMaps to store bucket lists sorted by ascending hashcode (that is, by ascending result of `computeHash(...)`). See for instance the private `put0` method:


```scala
  private[this] def put0(key: K, value: V, getOld: Boolean, hash: Int, idx: Int): Some[V] = {
    table(idx) match {
      case null =>
        table(idx) = new Node[K, V](key, hash, value, null)
      case old =>
        var prev: Node[K, V] = null
        var n = old
        while((n ne null) && n.hash <= hash) { // <-- keep traversing until we find the right spot for this hashcode
          if(n.hash == hash && key == n.key) {
            val old = n.value
            n.value = value
            return if(getOld) Some(old) else null
          }
          prev = n
          n = n.next
        }
        if(prev eq null) table(idx) = new Node(key, hash, value, old)
        else prev.next = new Node(key, hash, value, prev.next)
    }
    contentSize += 1
    null
  }
```

Reply:
Interesting. @szeiger what's the purpose of this sorting by hashcode? I can do the same here if necessary I guess

Reply:
It speeds up inserts and unsuccessful lookups because you don't have to scan the whole list. You won't see much of an effect for well distributed hashes with our standard load factor because buckets tend to be very small.

Reply:
I have replaced this code with a call to `put0` instead